uint8 address
uint8 quality_percents
